addons/hr_holidays/tests/test_allocations.py

```diff
@@ -89,3 +98,89 @@ def test_allocation_category(self):
+        allocation.write({'number_of_days_display': 14, 'number_of_days': 14})
+        self.assertEqual(allocation.number_of_days_display, 14)
+
+    def test_disallowed_change_allocation_due_to_taken_leaves(self):
+        allocation = self.env['hr.leave.allocation'].create({
+            'name': 'Initial Allocation',
+            'holiday_status_id': self.leave_type_paid.id,
+            'number_of_days': 10,
+            'employee_id': self.employee.id,
+            'date_from': date(2024, 1, 1),
+        })
+        allocation.action_validate()
+
+        leave_request = self.env['hr.leave'].create({
+            'name': 'Leave Request',
+            'holiday_status_id': self.leave_type_paid.id,
+            'request_date_from': date(2024, 1, 20),
+            'request_date_to': date(2024, 1, 29),
+            'number_of_days': 8,
+            'employee_id': self.employee.id,
+        })
+        leave_request.action_approve()
+
+        with self.assertRaises(ValidationError):
+            allocation.write({'number_of_days_display': 4, 'number_of_days': 4})
+
+    def test_disallowed_change_allocation_with_overlapping_allocations(self):
+        # Creating the first allocation
+        allocation_one = self.env['hr.leave.allocation'].create({
+            'name': 'First Allocation',
+            'holiday_status_id': self.leave_type_paid.id,
+            'number_of_days': 5,
+            'employee_id': self.employee.id,
+            'date_from': date(2024, 1, 1),
+            'date_to': date(2024, 1, 30),
+        })
+        allocation_one.action_validate()
+
+        # Creating the second overlapping allocation
+        allocation_two = self.env['hr.leave.allocation'].create({
+            'name': 'Second Half Allocation',
+            'holiday_status_id': self.leave_type_paid.id,
+            'number_of_days': 5,
+            'employee_id': self.employee.id,
+            'date_from': date(2024, 1, 20),
+            'date_to': date(2024, 2, 20),
+        })
+        allocation_two.action_validate()
+
+        # Creating a leave request consuming days from both allocations
+        leave_request = self.env['hr.leave'].create({
+            'name': 'Leave Request Spanning Allocations',
+            'holiday_status_id': self.leave_type_paid.id,
+            'request_date_from': date(2024, 1, 25),
+            'request_date_to': date(2024, 2, 5),
+            'employee_id': self.employee.id,
+        })
+        leave_request.action_approve()
+
+        update_amount = allocation_one.number_of_days + allocation_two.number_of_days - leave_request.number_of_days
+
+        with self.assertRaises(ValidationError):
+            allocation_one.write({'number_of_days_display': update_amount, 'number_of_days': update_amount})
+
+        allocation_one.write({'number_of_days_display': update_amount + 1, 'number_of_days': update_amount + 1})
+        self.assertEqual(allocation_one.number_of_days_display, update_amount + 1)
```

Comment:
This way of writing the values is a bit unclear. It works because of the values you chose, but it is not correct if we make the durations vary.
Also, I feel like the `assertEqual` at the end isn't meaningful; if the value isn't that one, an error will be raised because of the write not working.
```suggestion
        with self.assertRaises(ValidationError):
            allocation_one.write({'number_of_days': 2})

        allocation_one.write({'number_of_days': 3})
```